### DISCUSSION TITLE
How to update the GitLab token?

### DISCUSSION THREAD

**Discussion by NeXT405** [fluxcd/flux2#3825] [Q&A] [answered]
Hi there

I deleted a user yesterday from our GitLab instance. But the user was used for the ImageAutomation for Flux. I created a new user and generated a new Token for this user. 

How can I update the GitLab-Token? 

I tried to run the "bootstrap" command again with the  new environment variable. The command was successful, but the token in the cluster did not change. `export GITLAB_TOKEN=<your-token>`

```
flux bootstrap gitlab \
  --components-extra=image-reflector-controller,image-automation-controller \
  --hostname=gitlab.mycompany.ch \
  --token-auth \
  --owner=mycompany \
  --repository=k8config \
  --branch=develop
```

We use the below version in our develop cluster: 

```
flux: v0.37.0
helm-controller: v0.27.0
image-automation-controller: v0.27.0
image-reflector-controller: v0.23.0
kustomize-controller: v0.31.0
notification-controller: v0.29.0
source-controller: v0.32.1
```

Yeah, I know, I should update it... 

Can someone push me in the right direction and point me out the part in the documentation? 

Or, would it be safe to run the uninstall command and then the install command again? 

Thanks in advance.
Marcel

**Comment by makkes** [comment]
Bootstrapping again with a different token should in fact update the flux-system/flux-system Secret. Please make sure that the `GITLAB_TOKEN` environment variable contains the new value and is exported and then post the full output of your bootstrap run to debug this further.

**Comment by NeXT405** [comment, marked as answer]
Hi @makkes thanks for your response. I found that out few minutes later. 

Phew! That was a wild ride! 

I have found the problem. In the end, it was a combination of unfortunate circumstances. 

Here is my problem and the solution, maybe it will help someone in the future: 

1. yesterday evening I did a major update of our GitLab instance and deleted some, let's say, "unused" users. 
2. this morning I recreated the user for Flux and most likely assigned the wrong permissions. 
3. flux tried to log in to GitLab. -> Login Error
4. GitLab has banned the IPs. (good boy!)
5. GitLab returnd a lot of "503 Service Unavailable".

From now on it was not only a problem from the permissions but also from the blacklist. 

Solution: 

- Set permissions correctly for the new technical user in GitLab.
- Update all(!!) Secrets (in all namespaces where a `Kind: GitRepository` exists) with the correct user. 
- Remove IP from GitLab blacklist 
- Restart all Flux services

**GitLab actions:** 

**Whitelist the Node-IPs**

```bash
sudo nano /etc/gitlab/gitlab.rb

gitlab_rails['rack_attack_git_basic_auth'] = {
   'enabled' => true,
   'ip_whitelist' => ["<NODE-IP-1>","<NODE-IP-2>","<NODE-IP-3>","<NODE-IP-4>"],
   'maxretry' => 10,
   'findtime' => 60,
   'bantime' => 3600
 }
```

**Remove banned IPs:**

```bash
# Connect to redis
/opt/gitlab/embedded/bin/redis-cli -s /var/opt/gitlab/redis/redis.socket

# remove banned ips 
del cache:gitlab:rack::attack:allow2ban:ban:<ip>
```

But now we know that GitLab's security mechanisms work. :)
